src/main/java/Quaternion.java

```
@@ -0,0 +1,389 @@
+import org.jetbrains.annotations.*;
+
+import java.lang.invoke.VarHandle;
+import java.security.InvalidParameterException;
+
+/* Артём Олегович, здравствуйте!
+
+* Работа выполнена, даже чрезмерно и у меня есть к Вам вопрос,
+* касаемый принятых норм в написании классов:
+* сейчас у меня в одном классе мешанина из статических методов,
+* методов, возвращающие отредактированные копии, и
+* методов, преобразующие данные объекты. Скажите, насколько правильно
+* сохранять такую структуру класса и как правильно распределить всё,
+* если это мовитон?
+* */
+
+public class Quaternion {
+
+    private static final double equivalenceDelta = 1e-13;
+    private double w;
+    private Vector3 xyz;
+
+    public Quaternion() {
+        setW(0);
+        setVector(new Vector3());
+    }
+
+    public Quaternion(double w, double x, double y, double z) {
+        setW(w);
+        setVector(new Vector3(x, y, z));
+    }
+
+    public Quaternion(double w, Vector3 xyz) {
+        setW(w);
+        setVector(xyz);
+    }
+
+    public Quaternion(@NotNull Quaternion quaternion) {
+        setW(quaternion.getW());
+        setVector(quaternion.getVector());
+    }
+
+    public Quaternion(@NotNull Vector3 axis, double angle) {
+        setVector(axis.normalized().multiplied(Math.sin(angle / 2.0)));
+        setW(Math.cos(angle / 2.0));
+    }
+
+    public double getW() {
+        return w;
+    }
+
+    public double getX() { return getVector().getX(); }
+
+    public double getY() { return getVector().getY(); }
+
+    public double getZ() { return getVector().getZ(); }
+
+    public Vector3 getVector() {
+        return xyz;
+    }
+
+    public void setW(double w) {
+        validateInput(w);
```

Валидировать данные - это, безусловно, правильно и полезно. Вопрос вызывает другой момент - нужно ли это делать каждый раз, когда мы создаем новый объект. С одной стороны, ошибки лучше обнаруживать как можно раньше (подход [fail fast](https://en.wikipedia.org/wiki/Fail-fast)). С другой, это накладно в плане производительности - выполнять одну и ту же проверку много раз над одинаковыми данными. Кроме того, в таком случае практически любой метод кватерниона может выбросить исключение (мы этого не замечаем, потому что оно unchecked), и не всегда в вызывающем коде мы можем понять, как его обрабатывать. Подумайте, как лучше сделать и какой вариант более предпочтительный в данной задаче.

---

Я принял решение вообще не делать проверок. Вектор бесконечной длины может быть, а все операции с NaN, вернут NaN, поэтому будет понятно, что что-то идет не так. Для мануальной проверки "Все ли хорошо", просто перепишу метод validateInput на статичный. Думаю, это хороший вариант, т.к. все внутренние операции классов сделаны сейчас безопасно. Так что просто чтобы проследить за состоянием объекта можно будет проверять валидность данных.